resampling: Don't subscribe twice to the same source

leandro-lucarella-frequenz opened the pull request:
When receiving a subcription, we need to check if we are already handling a subscription with the same request parameters to avoid pushing the data twice to the same channel.

Fixes #115.

llucax committed:
Add some missing type hints to attributes

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

llucax committed:
resampling: Don't subcribe twice to the same source

When receiving a subcription, we need to check if we are already
handling a subscription with the same request parameters to avoid
pushing the data twice to the same channel.

Fixes #115.

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

leandro-lucarella-frequenz commented:
I will try to add a test to make sure we don´t regress.

shsms commented:
I can approve as soon as the black issues are fixed.  There will be tests from the logicalmeter anyway.

llucax committed:
mypy: Ignore async_solipsism.*

Newer async_solipsism versions provide some extra modules, so we ignore
those too.

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

leandro-lucarella-frequenz commented:
Updated to fix black issues and add a test. Enabling auto-merge.

llucax committed:
Fix wrong License header

Thanks fly to @Marenz for catching this.

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

ela-kotulska-frequenz commented on tests/actor/test_resampling.py:
```diff
@@ -118,19 +90,108 @@ async def test_component_metrics_resampling_actor(  # pylint: disable=too-many-s
     assert new_sample.value is None
+    assert new_sample.timestamp == _now()
+
+
+async def test_single_request(
```
Please use `@pytest.mark.asyncio` before async tests.

ela-kotulska-frequenz replied:
We don't need these wrapper, sorry :)

sahas-subramanian-frequenz reviewed (approved):
\o/

leandro-lucarella-frequenz commented:
After thinking about this a bit more, wondering how things could go after/if we add historical data, maybe I should have kept a set of requests instead of channels, as for historical data we might want to put more data in the same channel. But OTOH doing this will complicate a lot the use of the resampler, as whatever happens we should keep the pace of the resampling period. So better to address it when the problem comes and we have more clarity on how to solve it, o maybe we don't even use the same resampler for historical data than for live data.
